### PR TITLE
[JEP-237] Do not support shortening of HMAC code on FIPS mode

### PR DESCRIPTION
**Pull request by SujathaH**
HMACConfidentialKey supports an option to return trimmed HMAC code. Usage of trimmed code can cause issues when used from security context. This change is to throw an exception when callers try to get trimmed HMAC code on FIPS mode and also
this change assumes that Legacy Token is disabled on FIPS mode. There is a plan to remove support for Legacy Token and enhancement [JENKINS-71573](https://issues.jenkins.io/browse/JENKINS-71573) exists for same. Hence, Legacy Token flag  needs to be disabled on FIPS mode for this change to work successfully.

Verified across all the plugins and found there are no plugins which are using the truncated HMAC code.
The only impact is on Legacy Token in Jenkins Core

See [JEP-237](https://github.com/jenkinsci/jep/blob/master/jep/237/README.adoc).

### Testing done
Verified exception is thrown when HMAC Key is being created with length less than original length of HAMC code on FIPS mode
Verified trimmed code is returned when HMAC Key is being created with length less than 32 on non FIPS mode

### Proposed changelog entries

Prevent trimming HMAC codes (using `HAMCConfidentialKey`) when running in FIPS mode only.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@jtnord 

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/test/java/jenkins/security/HMACConfidentialKeyTest.java
Comment (jtnord) [inline core/src/test/java/jenkins/security/HMACConfidentialKeyTest.java:45]: we are chcking the lenght/size not code?

```java
    public void testMacWithShortenedLengthOnFips() {
```

or 
```java
    public void testTruncatedMacOnFips() {
```

File: core/src/test/java/jenkins/security/HMACConfidentialKeyTest.java
Comment (jtnord) [inline core/src/test/java/jenkins/security/HMACConfidentialKeyTest.java:56]: naming is hard - but the lenght/size is shortened not the code?
`testTruncatedMacOnNonFips` ?

**Review by jtnord — CHANGES_REQUESTED** [review]
the unit tests look like they will pass when they should fail.

Also please leave links to any plugins or code that will be impactec by this change in the summary.

File: test/src/test/java/hudson/security/HMACConfidentialKeyFIPSTest.java
Comment (jtnord) [inline test/src/test/java/hudson/security/HMACConfidentialKeyFIPSTest.java:29]: you are in FIPS mode here.
```suggestion
    public void testNonTruncatedMacOnFips() {
```

File: test/src/test/java/hudson/security/HMACConfidentialKeyFIPSTest.java
Comment (jtnord) [inline test/src/test/java/hudson/security/HMACConfidentialKeyFIPSTest.java:32]: please do not use assertTrue / assertFalse for any condition that is not **directly** a boolean but a computed boolean.
if it fails you do not get told the reason.
Hamcrest has a [matcher](https://hamcrest.org/JavaHamcrest/javadoc/2.2/org/hamcrest/text/MatchesPattern.html#matchesPattern-java.lang.String-) that will match a string against a pattern

File: test/src/test/java/hudson/security/HMACConfidentialKeyFIPSTest.java
Comment (jtnord) [inline test/src/test/java/hudson/security/HMACConfidentialKeyFIPSTest.java:25]:
```suggestion
        assertEquals("Supplied length can't be less than 32 on FIPS mode", iae.getMessage());
```

File: core/src/test/java/jenkins/security/HMACConfidentialKeyTest.java
Comment (jtnord) [inline core/src/test/java/jenkins/security/HMACConfidentialKeyTest.java:45]: please do not use assertTrue / assertFalse for any condition that is not **directly** a boolean but a computed boolean.
if it fails you do not get told the reason.
Hamcrest has a [matcher](https://hamcrest.org/JavaHamcrest/javadoc/2.2/org/hamcrest/text/MatchesPattern.html#matchesPattern-java.lang.String-) that will match a string against a pattern

**Review by jtnord — APPROVED** [review]

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/test/java/jenkins/security/HMACConfidentialKeyTest.java
Comment (DONGSIK2026KIM) [inline core/src/test/java/jenkins/security/HMACConfidentialKeyTest.java:46]: how about make a var as "Hello World"?

**Review by olamy — APPROVED** [review]

**Pull request merged**
